README.md
[![Build Status](https://img.shields.io/travis/Kaixhin/FGLab.svg)](https://travis-ci.org/Kaixhin/FGLab)
[![Dependency Status](https://img.shields.io/david/kaixhin/fglab.svg)](https://david-dm.org/Kaixhin/FGLab)
[![devDependency Status](https://img.shields.io/david/dev/kaixhin/fglab.svg)](https://david-dm.org/Kaixhin/FGLab#info=devDependencies)
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE.md)
[![Docker Pulls](https://img.shields.io/docker/pulls/kaixhin/fglab.svg)](https://hub.docker.com/r/kaixhin/fglab/)
[![Docker Stars](https://img.shields.io/docker/stars/kaixhin/fglab.svg)](https://hub.docker.com/r/kaixhin/fglab/)

# ![FGLab](public/images/fglab-logo.png)

### **Quickstart: https://kaixhin.github.io/FGLab/**

FGLab is a machine learning dashboard, designed to make prototyping experiments easier. Experiment details and results are sent to a database, which allows analytics to be performed after their completion. The server is FGLab, and the clients are [FGMachines](https://github.com/Kaixhin/FGMachine).

## Contents

- [Installation](#installation)
- [Overview](#overview)
- [Examples](#examples)
- [API](#api)

## Installation

FGLab tries to follow the [SemVer](http://semver.org/) standard whenever possible. Releases can be found [here](https://github.com/Kaixhin/FGLab/releases). There are 3 ways to run FGLab: Installing [locally](#option-1-local), via [Docker](#option-2-docker), or hosted on [Heroku](#option-3-heroku).

### Option 1: Local

1. Install [Node.js](https://nodejs.org/) from the website or your package manager.
1. Install [MongoDB](https://www.mongodb.org/) from the website or your package manager.
1. Make a database directory for MongoDB. For example, `mkdir -p <working directory>/db`.
1. Run the MongoDB daemon. From the previous example, run `mongod --dbpath <working directory>/db`.
1. Either clone this repository or download and extract a [zip](https://github.com/Kaixhin/FGLab/zipball/master)/[tar](https://github.com/Kaixhin/FGLab/tarball/master).
1. Move inside the FGLab folder.
1. Run `npm install`. `npm install` also runs `bower install` to install additional required packages.
1. FGLab requires a `.env` file in this directory. For most installations, it should be possible to copy [example.env](example.env) to `.env`, but it may require customisation for non-standard MongoDB ports, or setting a different port for FGLab. An alternative is to set the following environment variables:
  - MONGODB_URI (MongoDB database URI)
  - FGLAB_PORT (port)

Run `node lab` (or `npm start`) to start FGLab. You can now access the user interface from a browser on the current machine at `http://localhost:<FGLAB_PORT>`, where `<FGLAB_PORT>` is 5080 by default. Please read the [overview](#overview) to understand how FGLab and FGMachine cooperate - both are needed in order to run experiments. Afterwards, you should set up instances of [FGMachine](https://github.com/Kaixhin/FGMachine).

To update, run `npm run update`.

### Option 2: Docker

Start a [MongoDB container](https://hub.docker.com/_/mongo/) and link it to the [FGLab container](https://hub.docker.com/r/kaixhin/fglab/):

```sh
sudo docker run -d --name mongodb mongo
sudo docker run -d --name fglab --link mongodb:mongo -p 5080:5080 kaixhin/fglab
```

### Option 3: Heroku

The deploy button provisions a free dyno running FGLab on [Heroku](https://www.heroku.com), with a free 500MB MongoDB database from [MongoLab](https://mongolab.com/).

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

## Overview

FGLab is based on several classes of object. One begins with a *project*, which involves adjusting variables to achieve the desired results. In machine learning, these variables are *hyperparameters*, which are set for the project. In a more general setting, the variables are simply options, which may therefore include implementation-dependent details. A project will then comprise of a set of *experiments* derived from adjusting options.

### Projects

A project is created by uploading a [JSON](http://json.org/) schema. JSON is a human-readable data-interchange format that is widely used and has mature libraries available for most programming languages.

The JSON schema represents a map/associative array (without nesting), where the values are an object comprising of several fields:

- **type**:
  - `int`
  - `float`
  - `bool`
  - `string`
  - `enum`
- **default**: Default value
- **values**: An array of strings comprising the `enum`

See [mnist.json](test/mnist.json) as an example schema for a project. Each schema should be uploaded with the filename corresponding to the desired name for the project e.g. `mnist.json`.

Often it is hard to specify some options in advance e.g. the type or structure of the machine learning model. Sometimes code may change, which would influence the results. The `string` type can be used to address changing options and versioning manually e.g. `cnn.v2`.

This is stored by FGLab, and is used to construct a form which lets one choose options and submit an experiment to an available machine. The options are sent to your machine learning program via the FGMachine client. Your machine learning program then accepts the different fields via command-line options, the details of which are in the [FGMachine documentation](https://github.com/Kaixhin/FGMachine#projects). **Note that the `_id` field is reserved, as this will store the experiment ID as a `string`**.

FGMachine will spawn your machine learning program, which should produce output files to be sent from FGMachine to FGLab. The details of this is available in the [FGMachine documentation](https://github.com/Kaixhin/FGMachine#projects).

Grid and random search optimisers have also been implemented in FGLab, to allow searching over a range of hyperparameter space. Multiple string values are delimited by commas (`,`).

### Experiments

An experiment is one complete training and testing run with a specific set of options. Depending on the experiment it may be impossible to control for every source of randomness, so experiments with the same set of options will still be assigned unique IDs. Experiments have a unique ID, in addition to a project ID, a machine ID, the chosen options, the current status (running/success/fail), timestamps, results, and custom data; this provides a comprehensive record of the experiment as a whole.

The experiment page contains a "Logs" window, which uses WebSockets to display the experiment's `stdout` and `stderr` live. There is also an editable "Notes" text box that is automatically saved (at an interval of 0.5s), displaying on both the experiment page itself and the table of experiment results.

The current format for results is documented with [FGMachine](https://github.com/Kaixhin/FGMachine#experiments).

### Machines

A [FGMachine client](https://github.com/Kaixhin/FGMachine) registers itself with FGLab, providing hardware details as well as an address for interaction between FGLab and the machine. A machine (FGMachine) stores its own details, as well as a list of supported projects. Before a new experiment is chosen to be run, FGLab queries all machines in order to determine a machine with the capacity to run the experiment.

Note that machines are implementation-independent, and may well store their own (large) data on experiments, for example learnt parameters and logs. As mentioned before, these can be uploaded to FGLab's database.

## Examples

Examples utilising the range of abilities of FGLab/FGMachine can be found in the [examples folder](examples).

## API

The API is largely undocumented due to ongoing development introducing breaking changes. Ongoing documentation is available in [RAML](http://raml.org/): [api.raml](api.raml). The following are noted for convenience:

**Submit a new experiment with a set of options**
```
POST /api/v1/projects/{projectId}/experiment

e.g. curl -X POST -H "Content-Type: application/json" -d '{projectOptions}' http://{FGLab address}/api/v1/projects/{projectId}/experiment
```

If the project does not exist, returns `400 {"error": "Project ID <projectId> does not exist"}`. If the `projectOptions` are invalid, returns `400 {"error": "<validation message>"}`. If no machines are available to run the job, returns `501 {"error": "No machine capacity available"}`. If the machine fails to run the experiment for some reason, returns `500 {"error": "Experiment failed to run"}`. If successful, returns `201 {"_id": "<experimentId>"}`.

**Start a batch job with a list of option sets**
```
POST /api/v1/projects/{projectId}/batch?retry={retryTimeout (optional)}

e.g. curl -X POST -H "Content-Type: application/json" -d '[{projectOptions}]' http://{FGLab address}/api/v1/projects/{projectId}/batch?retry={retryTimeout (optional)}
```

The optional `retry` parameter specifies the maximum time in seconds to wait before trying to run a queued job again after capacity has been reached (the interval is randomly picked from a uniform distribution). If the project does not exist, returns `400 {"error": "Project ID <projectId> does not exist"}`. If any of the `projectOptions` are invalid, returns `400 {"error": "<validation message>"}` for the first set of options that are wrong. If successful, returns `201 {"status": "Started"}`. Future work aims to create a proper "optimiser" object that can be queried and have its work queue adjusted appropriately (hence differentiating it from a simple batch job queue).

**Register a webhook for an event**
```
POST /api/v1/webhooks

e.g. curl -X POST -H "Content-Type: application/json" -d '{webhookOptions}' http://{FGLab address}/api/v1/webhooks
```

`webhookOptions` expects the following options

```json
{
  "url": "<URL to POST to>",
  "objects": "<object collection to listen to (currently only 'experiments')>",
  "object_id": "<object ID>",
  "event": "<event to listen to (currently only 'started' or 'finished')>"
}
```

If a valid URL is not provided, returns `400 {"error": "Invalid or empty URL"}`. If a valid object collection is not provided, returns `400 {"error": "Object is not 'experiments'"}`. If a valid event is not provided, returns `400 {"error": "Event is not 'started' or 'finished'"}`. If an object ID is not provided, returns `400 {"error": "No object ID provided"}`. If successful, returns `201 {"status": "Registered", "options": <webhookOptions>"}`. When the event occurs, the JSON data used to register the webhook is returned.


CHANGELOG.md
<a name="0.6.0"></a>
# [0.6.0](https://github.com/Kaixhin/FGLab/compare/0.6.0...v0.6.0) (2016-01-11)


### Bug Fixes

* fix random search evaluations number ([8208e36](https://github.com/Kaixhin/FGLab/commit/8208e36))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/Kaixhin/FGLab/compare/0.5.0...v0.5.0) (2015-12-11)


### Bug Fixes

* add all score properties across experiments ([30cc3be](https://github.com/Kaixhin/FGLab/commit/30cc3be))
* add ws disconnect err handler ([3acd9ba](https://github.com/Kaixhin/FGLab/commit/3acd9ba))
* correct metaparam IDs for hyperparam form ([e827471](https://github.com/Kaixhin/FGLab/commit/e827471)), closes [#13](https://github.com/Kaixhin/FGLab/issues/13)
* fix status indicators for multiple machines ([2631cfa](https://github.com/Kaixhin/FGLab/commit/2631cfa))
* revert to jQuery 3 (for then) with Bootstrap 4 alpha 1 ([38d843f](https://github.com/Kaixhin/FGLab/commit/38d843f))
* satisfy mongoskin-mongodb peer dependency ([9c9e7d0](https://github.com/Kaixhin/FGLab/commit/9c9e7d0))

### Features

* add autosaving notes field per experiment ([8e156e1](https://github.com/Kaixhin/FGLab/commit/8e156e1))
* add live logging ([4c1084c](https://github.com/Kaixhin/FGLab/commit/4c1084c)), closes [#9](https://github.com/Kaixhin/FGLab/issues/9)
* option selection for enums/bools in optimisation ([e8c63ef](https://github.com/Kaixhin/FGLab/commit/e8c63ef))

### Performance Improvements

* display server error for uploading project schema ([a17676c](https://github.com/Kaixhin/FGLab/commit/a17676c))
* upgrade to Bootstrap v4 alpha 2 ([c120525](https://github.com/Kaixhin/FGLab/commit/c120525))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/Kaixhin/FGLab/compare/0.4.0...v0.4.0) (2015-11-01)


### Bug Fixes

* allow bower to install with root ([974c99e](https://github.com/Kaixhin/FGLab/commit/974c99e))
* delete _evaluations hidden property on grid search ([4731605](https://github.com/Kaixhin/FGLab/commit/4731605))
* save multiple objects in GridFS and download binaries ([1186de5](https://github.com/Kaixhin/FGLab/commit/1186de5))

### Features

* add machine live indicator ([39be32a](https://github.com/Kaixhin/FGLab/commit/39be32a))
* add machine liveness indicator to home ([825897a](https://github.com/Kaixhin/FGLab/commit/825897a))
* add server status indicator ([e46c09f](https://github.com/Kaixhin/FGLab/commit/e46c09f))
* add update script ([92e953a](https://github.com/Kaixhin/FGLab/commit/92e953a))
* display uploaded images ([a706419](https://github.com/Kaixhin/FGLab/commit/a706419))
* display videos ([45363de](https://github.com/Kaixhin/FGLab/commit/45363de))
* list machines' supported projects ([9a02ce1](https://github.com/Kaixhin/FGLab/commit/9a02ce1))

### Performance Improvements

* add project ID to homepage ([2b46119](https://github.com/Kaixhin/FGLab/commit/2b46119))
* list options and sort filenames ([adfae66](https://github.com/Kaixhin/FGLab/commit/adfae66))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/Kaixhin/FGLab/compare/0.3.0...v0.3.0) (2015-10-16)


### Bug Fixes

* delete files when experiment deleted ([4312e8c](https://github.com/Kaixhin/FGLab/commit/4312e8c))
* downsample different axis types for C3 ([864ba4b](https://github.com/Kaixhin/FGLab/commit/864ba4b))

### Features

* accept C3 API for charts ([cc97ff7](https://github.com/Kaixhin/FGLab/commit/cc97ff7))



<a name="0.2.7"></a>
## [0.2.7](https://github.com/Kaixhin/FGLab/compare/0.2.7...v0.2.7) (2015-10-14)


### Bug Fixes

* update and add missing URL regex condition ([a999f28](https://github.com/Kaixhin/FGLab/commit/a999f28))

### Features

* add Moment.js for parsing dates ([22d707e](https://github.com/Kaixhin/FGLab/commit/22d707e))
* add automated bump parser based on commit messages ([58b1493](https://github.com/Kaixhin/FGLab/commit/58b1493))
* use general _scores object ([e4f58b6](https://github.com/Kaixhin/FGLab/commit/e4f58b6))



<a name="0.2.7"></a>
## [0.2.7](https://github.com/Kaixhin/FGLab/compare/0.2.6...v0.2.7) (2015-10-14)




<a name="0.2.6"></a>
## 0.2.6 (2015-10-14)


### Features

* add gulp release task with contributing doc ([a477a43](https://github.com/Kaixhin/FGLab/commit/a477a43))





lab.js
/* Modules */
require("./env"); // Load configuration variables
var http = require("http");
var path = require("path");
var EventEmitter = require("events").EventEmitter;
var mediator = new EventEmitter();
var _ = require("lodash");
var express = require("express");
var bodyParser = require("body-parser");
var multer = require("multer");
var compression = require("compression");
var favicon = require("serve-favicon");
var morgan = require("morgan");
var rp = require("request-promise");
var Promise = require("bluebird");
var WebSocketServer = require("ws").Server;
var db = require("./db").db;

/* App instantiation */
var app = express();
var jsonParser = bodyParser.json({limit: '100mb'}); // Parses application/json
var upload = multer(); // Store files in memory as Buffer objects
app.use(compression()); // Compress all Express requests
app.use(favicon(path.join(__dirname, "public/favicon.ico"))); // Deal with favicon requests
app.use(express.static(path.join(__dirname, "public"), {index: false, maxAge: '1d'})); // Static directory
app.use("/bower_components", express.static(path.join(__dirname, "bower_components"), {index: false, maxAge: '1d'})); // Bower components
app.set("view engine", "jade"); // Jade template engine
app.use(morgan("tiny")); // Log requests

/* API */

// Registers webhooks
app.post("/api/v1/webhooks", jsonParser, (req, res) => {
  // Parse webhook request
  var webhook = req.body;
  var url = webhook.url;
  var objects = webhook.objects;
  var event = webhook.event;
  var objId = webhook.object_id;

  // Use John Gruber's URL regex
  var urlRegEx = /\b((?:[a-z][\w-]+:(?:\/{1,3}|[a-z0-9%])|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}\/)\S+(?:[^\s`!\[\]{};:'".,?«»“”‘’]))/ig;
  // Validate
  if (!url || !urlRegEx.test(url)) {
    res.status(400);
    return res.send({error: "Invalid or empty URL"});
  } else if (objects !== "experiments") {
    res.status(400);
    return res.send({error: "Object is not 'experiments'"});
  } else if (event !== "started" && event !== "finished") {
    res.status(400);
    return res.send({error: "Event is not 'started' or 'finished'"});
  } else if (!objId) {
    // TODO Check object exists
    res.status(400);
    return res.send({error: "No object ID provided"});
  }

  // Register with mediator
  mediator.once(objects + ":" + objId + ":" + event, () => {
    // Send webhook response
    rp({uri: webhook.url, method: "POST", json: webhook, gzip: true})
    .catch(() => {}); // Ignore failures from missing webhooks
  });
  res.status(201);
  return res.send({status: "Registered", options: webhook});
});

// Downloads file
app.get("/api/v1/files/:id", (req, res, next) => {
  // Open file
  var gfs = new db.GridStore(db, db.toObjectID(req.params.id), "r", {promiseLibrary: Promise});
  gfs.open((err, gfs) => {
    // Set read head pointer to beginning of file
    gfs.seek(0, () => {
      // Read entire file
      gfs.read()
      .then((file) => {
        res.setHeader("Content-Disposition", "attachment; filename=" + gfs.filename); // Set as download
        res.setHeader("Content-Type", gfs.metadata.contentType); // MIME Type
        res.send(file); // Send file
      })
      .catch((err) => {
        next(err); 
      });
    });
  });
});

// Get collection for all API db-based endpoints
app.param("collection", (req, res, next, collection) => {
  req.collection = db.collection(collection);
  return next();
});

// Return all entries
app.get("/api/v1/:collection", (req, res, next) => {
  req.collection.find({}).toArrayAsync()
  .then((results) => {
    res.send(results);
  })
  .catch((err) => {
    next(err);
  });
});

// Create new entry
app.post("/api/v1/:collection", jsonParser, (req, res, next) => {
  req.collection.insertAsync(req.body, {})
  .then((result) => {
    res.status(201);
    res.send(result.ops[0]);
  })
  .catch((err) => {
    next(err);
  });
});

// Return single entry
app.get("/api/v1/:collection/:id", (req, res, next) => {
  req.collection.findByIdAsync(req.params.id)
  .then((result) => {
    res.send(result);
  })
  .catch((err) => {
    next(err);
  });
});

// Update existing entry
app.put("/api/v1/:collection/:id", jsonParser, (req, res, next) => {
  delete req.body._id; // Delete ID (will not update otherwise)
  req.collection.updateByIdAsync(req.params.id, {$set: req.body})
  .then((result) => {
    // Update returns the count of affected objects
    res.send((result === 1) ? {msg: "success"} : {msg: "error"});
  })
  .catch((err) => {
    next(err);
  });
});

// Delete existing entry
app.delete("/api/v1/:collection/:id", (req, res, next) => {
  req.collection.removeByIdAsync(req.params.id)
  .then((result) => {
    // Remove returns the count of affected objects
    res.send((result === 1) ? {msg: "success"} : {msg: "error"});
  })
  .catch((err) => {
    next(err);
  });
});

// Return all experiments for a project
app.get("/api/v1/projects/:id/experiments", (req, res, next) => {
  db.experiments.find({_project_id: db.toObjectID(req.params.id)}).toArrayAsync() // Get experiments for project
  .then((result) => {
    res.send(result);
  })
  .catch((err) => {
    next(err);
  });
});

// Delete all experiments for a project
app.delete("/api/v1/projects/:id/experiments", (req, res, next) => {
  db.experiments.removeAsync({_project_id: db.toObjectID(req.params.id)}) // Get experiments for project
  .then((result) => {
    res.send(result);
  })
  .catch((err) => {
    next(err);
  });
});

// TODO Consider renaming API endpoint
// Constructs a project from an uploaded .json file
app.post("/api/v1/projects/schema", upload.single("schema"), (req, res, next) => {
  // Extract file name
  var name = req.file.originalname.replace(".json", "");
  // Extract .json as object
  var schema = JSON.parse(req.file.buffer.toString());
  // Store
  db.projects.insertAsync({name: name, schema: schema}, {})
  .then((result) => {
    res.send(result);
  })
  .catch((err) => {
    next(err);
  });
});

var optionChecker = (schema, obj) => {
  for (var prop in schema) {
    var schemaField = schema[prop];
    var val = obj[prop];
    // Check field exists
    if (val === undefined) {
      return {error: "Field " + prop + " missing"};
    }
    // Check field is valid
    var type = schemaField.type;
    // int float bool string enum
    if (type === "int") {
      if (isNaN(val) ||  val % 1 !== 0) {
        return {error: "Field " + prop + " of type " + type + " is invalid"};
      }
    } else if (type === "float") {
      if (isNaN(val)) {
        return {error: "Field " + prop + " of type " + type + " is invalid"};
      }
    } else if (type === "bool") {
      if (val !== true && val !== false) {
        return {error: "Field " + prop + " of type " + type + " is invalid"};
      }
    } else if (type === "string") {
      if (val.length === 0) {
        return {error: "Field " + prop + " of type " + type + " is invalid"};
      }
    } else if (type === "enum") {
      if (schemaField.values.indexOf(val) === -1) {
        return {error: "Field " + prop + " of type " + type + " is invalid"};
      }
    }
  }
  return {success: "Options validated"};
};

var submitJob = (projId, options) => {
  return new Promise((resolve, reject) => {
    db.machines.find({}, {address: 1}).toArrayAsync() // Get machine hostnames
    .then((machines) => {
      var macsP = Array(machines.length);
      // Check machine capacities
      for (var i = 0; i < machines.length; i++) {
        macsP[i] = rp({uri: machines[i].address + "/projects/" + projId + "/capacity", method: "GET", data: null});
      }

      // Loop over reponses
      Promise.any(macsP)
      // First machine with capacity, so use
      .then((availableMac) => {
        availableMac = JSON.parse(availableMac);

        // Create experiment
        db.experiments.insertAsync({_options: options, _project_id: db.toObjectID(projId), _machine_id: db.toObjectID(availableMac._id), _files: [], _status: "running"}, {})
        .then((exp) => {
          options._id = exp.ops[0]._id.toString(); // Add experiment ID to sent options
          // Send project
          rp({uri: availableMac.address + "/projects/" + projId, method: "POST", json: options, gzip: true})
          .then((body) => {
            resolve(body);
          })
          .catch(() => {
            db.experiments.removeByIdAsync(exp.ops[0]._id); // Delete failed experiment
            reject({error: "Experiment failed to run"});
          });
        })
        .catch((err) => {
          reject(err);
        });
      })
      // No machines responded, therefore fail
      .catch(() => {
        reject({error: "No machine capacity available"});
      });
    })
    .catch((err) => {
      reject(err);
    });
  });
};

// Constructs an experiment
app.post("/api/v1/projects/:id/experiment", jsonParser, (req, res, next) => {
  var projId = req.params.id;
  // Check project actually exists
  db.projects.findByIdAsync(projId, {schema: 1})
  .then((project) => {
    if (project === null) {
      res.status(400);
      res.send({error: "Project ID " + projId + " does not exist"});
    } else {
      var obj = req.body;

      // Validate
      var validation = optionChecker(project.schema, obj);
      if (validation.error) {
        res.status(400);
        res.send(validation);
      } else {
        submitJob(projId, obj)
        .then((resp) => {
          res.status(201);
          res.send(resp);
        })
        .catch((err) => {
          // TODO Check comprehensiveness of error catching
          if (err.error === "No machine capacity available") {
            res.status(501);
            res.send(err);
          } else if (err.error === "Experiment failed to run") {
            res.status(500);
            res.send(err);
          } else {
            next(err);
          }
        });
      }
    }
  })
  .catch((err) => {
    next(err);
  });
});

// Submit job with retry
var submitJobRetry = function(projId, options, retryT) {
  submitJob(projId, options)
  .then(() => {
    // TODO Keep track of batch jobs
  })
  .catch(() => {
    // Retry in a random 1s interval
    setTimeout(() => {
      submitJobRetry(projId, options, retryT);
    }, 1000*retryT*Math.random());
  });
};

// Constructs a batch job from a list of options
app.post("/api/v1/projects/:id/batch", jsonParser, (req, res, next) => {
  var projId = req.params.id;
  var retryTimeout = parseInt(req.query.retry);
  // Set default as an hour
  if (isNaN(retryTimeout) || retryTimeout <= 0) {
    retryTimeout = 60*60;
  }
  // Check project actually exists
  db.projects.findByIdAsync(projId, {schema: 1})
  .then((project) => {
    if (project === null) {
      res.status(400);
      res.send({error: "Project ID " + projId + " does not exist"});
    } else {
      var expList = req.body;
      // Validate
      var validationList = [];
      for (var i = 0; i < expList.length; i++) {
       var validation = optionChecker(project.schema, expList[i]);
        if (validation.error) {
          validationList.push(validation);
        }
      }
      if (validationList.length > 0) {
        res.status(400);
        res.send(validationList[0]); // Send first validation error       
      } else {
        // Loop over jobs
        for (var j = 0; j < expList.length; j++) {
          submitJobRetry(projId, expList[j], retryTimeout);
        }
        res.send({status: "Started"});
      }
    }
  })
  .catch((err) => {
    next(err);
  });
});

// Adds started time to experiment
app.put("/api/v1/experiments/:id/started", (req, res, next) => {
  mediator.emit("experiments:" + req.params.id + ":started"); // Emit event

  db.experiments.updateByIdAsync(req.params.id, {$set: {_started: new Date()}})
  .then((result) => {
    // Update returns the count of affected objects
    res.send((result === 1) ? {msg: "success"} : {msg: "error"});
  })
  .catch((err) => {
    next(err);
  });
});

// Adds finished time to experiment
app.put("/api/v1/experiments/:id/finished", (req, res, next) => {
  mediator.emit("experiments:" + req.params.id + ":finished"); // Emit event

  db.experiments.updateByIdAsync(req.params.id, {$set: {_finished: new Date()}})
  .then((result) => {
    // Update returns the count of affected objects
    res.send((result === 1) ? {msg: "success"} : {msg: "error"});
  })
  .catch((err) => {
    next(err);
  });
});

// Processess files for an experiment
app.put("/api/v1/experiments/:id/files", upload.array("_files"), (req, res, next) => {
  delete req.body._id; // Delete ID (will not update otherwise)
  var filesP = Array(req.files.length);

  var saveGFSFile = function(fileObj) {
    var fileId = new db.ObjectID(); // Create file ID
    // Open new file
    var gfs = new db.GridStore(db, fileId, fileObj.originalname, "w", {metadata: {contentType: fileObj.mimetype}, promiseLibrary: Promise});
    gfs.open((err, gfs) => {
      if (err) {
        console.log(err);
      } else {
        // Write from buffer and flush to db
        gfs.write(fileObj.buffer, true)
        .then((gfs) => {
          // Save file reference
          filesP[i] = db.experiments.updateByIdAsync(req.params.id, {$push: {_files: {_id: gfs.fileId, filename: gfs.filename, mimetype: gfs.metadata.contentType}}});
        })
        .catch((err) => {
          console.log(err);
        });
      }
    });
  };

  for (var i = 0; i < req.files.length; i++) {
    saveGFSFile(req.files[i]); // Save file in function closure
  }

  // Check file promises
  Promise.all(filesP)
  .then(() => {
    res.send({message: "Files uploaded"});
  })
  .catch((err) => {
    next(err);
  });
});

// Delete all files for an experiment
app.delete("/api/v1/experiments/:id/files", (req, res, next) => {
  db.experiments.findByIdAsync(req.params.id, {_files: 1})
  .then((experiment) => {
    var filesP = Array(experiment._files.length);

    for (var i = 0; i < experiment._files.length; i++) {
      var gfs = new db.GridStore(db, experiment._files[i]._id, "w", {promiseLibrary: Promise});
      filesP[i] = gfs.unlinkAsync();
    }

    // Check file promises
    Promise.all(filesP)
    .then(() => {
      res.send({message: "Files deleted"});
    })
    .catch((err) => {
      console.log(err);
      next(err);
    });
  })
  .catch((err) => {
    console.log(err);
    next(err);
  });
});

// Delete all files for a project
app.delete("/api/v1/projects/:id/experiments/files", (req, res, next) => {
  db.experiments.find({_project_id: db.toObjectID(req.params.id)}).toArrayAsync() // Get experiments for project
  .then((experiments) => {
    var numFiles = 0;
    for (var i = 0; i < experiments.length; i++) {
      numFiles += experiments[i]._files.length;
    }
    var filesP = Array(numFiles);

    // Loop over experiments
    var fileIndex = 0;
    for (var j = 0; j < experiments.length; j++) {
      var experiment = experiments[j];
      // Loop over files
      for (var k = 0; k < experiment._files.length; k++) {
        var gfs = new db.GridStore(db, experiment._files[k]._id, "w", {promiseLibrary: Promise});
        filesP[fileIndex++] = gfs.unlinkAsync();
      }
    }

    // Check file promises
    Promise.all(filesP)
    .then(() => {
      res.send({message: "Files deleted"});
    })
    .catch((err) => {
      next(err);
    });
  })
  .catch((err) => {
    next(err);
  });
});

// Registers machine projects
app.post("/api/v1/machines/:id/projects", jsonParser, (req, res, next) => {
  db.machines.findByIdAsync(req.params.id)
  .then((result) => {
    // Fail if machine does not exist
    if (result === null) {
      res.status(404);
      return res.send({error: "Machine ID " + req.params.id + " does not exist"});
    }
    // Register projects otherwise
    db.machines.updateByIdAsync(req.params.id, {$set: req.body})
    .then((result) => {
      // Update returns the count of affected objects
      res.send((result === 1) ? {msg: "success"} : {msg: "error"});
    })
    .catch((err) => {
      next(err);
    });
  })
  .catch((err) => {
    next(err);
  });
});

/* Rendering Routes */

// List projects and machines on homepage
app.get("/", (req, res, next) => {
  var projP = db.projects.find({}, {name: 1}).sort({name: 1}).toArrayAsync(); // Get project names
  var macP = db.machines.find({}, {address: 1, hostname: 1}).sort({hostname: 1}).toArrayAsync(); // Get machine addresses and hostnames
  Promise.all([projP, macP])
  .then((results) => {
    return res.render("index", {projects: results[0], machines: results[1]});
  })
  .catch((err) => {
    return next(err);
  });
});

// Project page (new experiment)
app.get("/projects/:id", (req, res, next) => {
  db.projects.findByIdAsync(req.params.id)
  .then((result) => {
    res.render("project", {project: result});
  })
  .catch((err) => {
    next(err);
  });
});

// Project page (optimisation)
app.get("/projects/:id/optimisation", (req, res, next) => {
  db.projects.findByIdAsync(req.params.id)
  .then((result) => {
    res.render("optimisation", {project: result});
  })
  .catch((err) => {
    next(err);
  });
});

// Project page (experiments)
app.get("/projects/:id/experiments", (req, res, next) => {
  var projP = db.projects.findByIdAsync(req.params.id);
  var expP = db.experiments.find({_project_id: db.toObjectID(req.params.id)}, {_scores: 1, _status: 1, _options: 1, _started: 1, _finished: 1, _notes: 1}).toArrayAsync();
  Promise.all([projP, expP])
  .then((results) => {
    res.render("experiments", {project: results[0], experiments: results[1]});
  })
  .catch((err) => {
    next(err);
  });
});

// Machine page
app.get("/machines/:id", (req, res, next) => {
  db.machines.findByIdAsync(req.params.id)
  .then((mac) => {
    var projKeys = _.keys(mac.projects); // Extract project IDs
    projKeys = _.map(projKeys, db.toObjectID); // Map to MongoDB IDs
    db.projects.find({_id: {$in: projKeys}}, {name: 1}).sort({name: 1}).toArrayAsync()
    .then((projects) => {
      // Return only projects existing in FGLab
      res.render("machine", {machine: mac, projects: projects});
    })
    .catch((err) => {
      next(err);
    });
  })
  .catch((err) => {
    next(err);
  });
});

// Experiment page
app.get("/experiments/:id", (req, res, next) => {
  db.experiments.findByIdAsync(req.params.id)
  .then((result) => {
    var projP = db.projects.findByIdAsync(result._project_id, {name: 1}); // Find project name
    var macP = db.machines.findByIdAsync(result._machine_id, {hostname: 1, address: 1}); // Find machine hostname and address
    Promise.all([projP, macP]) 
    .then((results) => {
      res.render("experiment", {experiment: result, project: results[0], machine: results[1]});
    })
    .catch((err) => {
      next(err);
    });
  })
  .catch((err) => {
    next(err);
  });
});


/* Errors */
// Error handler
app.use((err, req, res, next) => {
  if (res.headersSent) {
    return next(err); // Delegate to Express' default error handling
  }
  res.status(500);
  res.send("Error: " + err);
});

/* HTTP server */
var server = http.createServer(app); // Create HTTP server
if (!process.env.FGLAB_PORT) {
  console.log("Error: No port specified");
  process.exit(1);
} else {
  // Listen for connections
  server.listen(process.env.FGLAB_PORT, () => {
    console.log("Server listening on port " + process.env.FGLAB_PORT);
  });
}

/* WebSocket server */
// Add websocket server
var wss = new WebSocketServer({server: server});
// Catches errors to prevent FGMachine crashing if browser disconnect undetected
var wsErrHandler = function() {};

// Call on connection from new client
wss.on("connection", (ws) => {
  // Print received messages
  ws.on("message", (message) => {
    console.log(message);
  });

  // Perform clean up if necessary
  ws.on("close", () => {
    //console.log("Client closed connection");
  });
});
